design/one-pager-strongly-typed-class.md

```diff
@@ -0,0 +1,218 @@
+
+## Goals
+
+* Enable schema documentation and validation for resource classes.
+* Continue to support configuring a default resource class to be set upon resource claims that do not specify a class.
+* Assume that the universe of possible default resource class kinds for a specific claim kind will grow arbitrarily, so a default class controller must be able to be made aware of new resource class kinds. This goal will become specifically important as we begin to [separate providers](https://github.com/crossplaneio/crossplane/issues/531) from the core Crossplane project in the form of infrastructure stacks. 
+
+## Proposal
+
+Unfortunately, strongly typed resource classes make it much more difficult to identify default resource classes. Currently, there is one default class controller per claim kind that watches for creation of their specified claim kind, then lists objects of `kind: ResourceClass` that are labeled as default for the claim kind. With an unknown number of possible kinds of resource classes that could be installed and serve as default for a claim `kind`, a default class controller can no longer list a single `ResourceClass` kind.
+
+One solution would instead be to map resource classes one-to-one with claim kinds (i.e. `MySQLInstanceClass`), but that results in the same issue of varying required parameters that is currently being experienced.
+
+In order to satisfy both the requirements of strong typing and default resource classes, a new custom resource kind, `Defaulter`, can be implemented to specify when a specific resource class kind can serve as default for a given claim kind. The workflow would proceed as follows:
+
+1. An admin installs Crossplane, which specifies the portable claim kinds that are available as part of the core Crossplane.
+2. An admin installs an infrastructure stack (i.e. a set of CRDs and controllers for a provider's managed service offerings).
+3. An admin creates an instance of an installed resource class CRD (i.e. `kind: RDSInstanceClass`). It could look as follows:
+
+```yaml
+apiVersion: database.aws.crossplane.io/v1alpha1
+kind: RDSInstanceClass
+metadata:
+  name: standard-mysql
+  namespace: crossplane-system
+spec:
+  class: db.t2.small
+  masterUsername: masteruser
+  securityGroups: "sg-ab1cdefg,sg-05adsfkaj1ksdjak"
+  size: "20"
+  provisioner: rdsinstance.database.aws.crossplane.io/v1alpha1
+providerRef:
+  name: aws-provider
+reclaimPolicy: Delete
+```
+
+4. An admin creates a new `Defaulter` object, specifying that the previously created resource class (`kind: RDSInstanceClass`) should serve as default for a given claim kind `mysqlinstance.storage.crossplane.io`. A `Defaulter` could look as follows:
+
+```yaml
+apiVersion: core.crossplane.io/v1alpha1
+kind: Defaulter
+metadata:
+  name: mysql-aws-defaulter
+  namespace: crossplane-system
+  labels:
+    defaultFor: mysqlinstance.storage.crossplane.io
```

Comment:
The only downside I see to this path is with the creation of `Defaulter` objects. However, because you have to specify the claim kind in the `Defaulter` anyways, I don't think it is a larger burden to create an object of kind `MySQLInstanceDefaulter` than of just `Defaulter`.

If we are to move towards defaulting other claim fields with this same object, are you imagining that these fields will also be object references? If not, will the structure of the `MySQLInstancePolicy` object have other components like `defaultClass` below, and a `Policy` may define one or more type of default behavior?

Right now, I am imagining the following:
```yaml
apiVersion: core.crossplane.io/v1alpha1
kind: MySQLInstancePolicy
metadata:
  name: mysql-aws-
  namespace: crossplane-system
defaultClass:
  kind: rdsinstanceclass.database.aws.crossplane.io/v1alpha1
  name: standard-mysql
  namespace: crossplane-system
```

Reply:
Yeah, that's what I was thinking. The only cases I have in mind would be setting default references. So something like:

```yaml
apiVersion: core.crossplane.io/v1alpha1
kind: MySQLInstancePolicy
metadata:
  name: mysql-aws-policy
  namespace: crossplane-system
defaultClassRef:
  kind: rdsinstanceclass.database.aws.crossplane.io/v1alpha1
  name: standard-mysql
  namespace: crossplane-system
defaultProviderRef:
  kind: provider.aws.crossplane.io/v1alpha1
  name: experimentalaccount
  namespace: crossplane-system
defaultSomethingElseRef:
  kind: coolthing.aws.crossplane.io/v1alpha1
  name: somethingcool
  namespace: crossplane-system
```

Reply:
To start, would we just want to design the CRD's with `defaultClassRef` and then add potential additional support in the future?

Reply:
> To start, would we just want to design the CRD's with defaultClassRef

Yeah, let's just cover this case for now.